### DISCUSSION TITLE
How to create my own Terminal Emulator

### DISCUSSION THREAD

**Discussion by a-usr** [microsoft/terminal#16348] [Q&A]
I am thinking about creating my own Terminal Emulator of sorts. However although the internet is full of the most random things, I find it surprisingly difficult to find any information on creating a Terminal Emulator for Windows (or in general).

 This (and the suggestion @lhecker gave me [here](https://github.com/MicrosoftDocs/Console-Docs/issues/303#issuecomment-1820931747)) gave me the idea to just direct my questions to the devs and community of my **favourite** Terminal Emulator.
 Aditionally this Discussion can also serve as some sort of guide to others (given that every important question is answered)

To start things off, here is my first Question: Currently the ony implementations of Terminal Emulators i have seen the source code of are written in c++. However, while i do know a bit of c++ myself, i personally prefer C#. Would it be possible or smart to create a Terminal Emulator in C#, or is it not possible because of performance and/or language restrictions?

**Comment by a-usr** [comment]
Ok, here is what i have compiled so far:
To connect a process to the Terminal Emulator (from this point I'll just call it T Emu for shortness's sake) the T Emu has to start it with certain flags. In c++ this is done by supplying `CreateProcess` with a specially crafted `StartupInfo`, wich has the `STARTF_USESTDHANDLES` flag and whose `hStdInput`, `hStdOutput` and `hStdError` attributes contain handles wich may or may not is being created by a function called `nt::console::handle` (https://github.com/o-sdn-o/vtm does it like that ([here](https://github.com/o-sdn-o/vtm/blob/d38bf03c2e74124d72c31de5c7623f96864e80fa/src/netxs/desktopio/consrv.hpp#L55)) however i cant find ANY documentation about this function, wich, even though the mentioned file doesn't `#include` any other files, leads me to believe that that function is either windows-specific and undocumented (wich i think would be rather improbable(or is the word "unprobable"?)) or custom defined...)

**Reply by lhecker** [reply]
Try setting a breakpoint here: https://github.com/microsoft/terminal/blob/376737e54afc301ced37d5416bc836eadbc611d2/src/server/DeviceHandle.cpp#L58

**Reply by a-usr** [reply]
I am currently having problems compiling Host.EXE (`OpenCosole.exe` missing). I'll look into it tomorrow.

**Comment by lhecker** [comment]
> To start things off, here is my first Question: Currently the ony implementations of Terminal Emulators i have seen the source code of are written in c++. However, while i do know a bit of c++ myself, i personally prefer C#. Would it be possible or smart to create a Terminal Emulator in C#, or is it not possible because of performance and/or language restrictions?

Yeah, you can use C# to implement the console APIs. Since C# can call native APIs, including `LoadLibrary` and `GetProcAddress` this should not pose any problems. Performance, apart from the startup performance, will likely not be a big problem for two reasons:
* The console API is effectively just a byte stream that you parse, similar to parsing data on TCP sockets. You won't be bottlenecked by C#'s slightly slower execution speed there.
* Try benchmarking terminals on Linux and you'll find that the fastest terminal in almost every category is [xterm.js](https://github.com/xtermjs/xterm.js/), a language that doesn't even have proper native byte slices. While your upper performance ceiling will be limited by the choice of your language, experience will carry you further than most others. I estimate that C# offers you a performance ceiling that's at least >20x over the performance of today's Windows Terminal. So if you're fine with that, you'll be fine with C#. 🙂

**Comment by lhecker** [comment]
BTW, I've been assuming that you already knew this and were just looking for writing a terminal emulator from scratch, but...

You do know that Windows does offer a simplified translation layer between console clients and VT sequences right? You can find more information here: https://learn.microsoft.com/en-us/windows/console/creating-a-pseudoconsole-session
It's not particularly fast, but that way you at least only need to implement a VT parser, text buffer and text renderer.

**Reply by a-usr** [reply]
Well, TBH while I did read all the articles in the "About Consoles" section, I only read a few of the articles in the API references and then came to the conclusion that all of the API described there is for the consumer side aka the Applications requiring a T Emu. 
Now of course I realize that that was a mistake. Thank you! 

However I am still curious how I would accomplish this myself, and I feel like this discussion wouldn't be complete without me trying to attempt to create a working T Emu without using the translaton layer provided by Windows as my intention for this discussion is not only to get help, but also to provide some kind of walkthrough for others looking to do something similar once I finished this.

**Comment by a-usr** [comment]
Do i have to implement line wrapping myself, or is that something the consumer side already has to do?
